ForumProject/communications/static/room.js

```diff
@@ -0,0 +1,105 @@
+    if (document.querySelector("option[value='" + value + "']")) return;
+    let newOption = document.createElement("option");
+    newOption.value = value;
+    newOption.innerHTML = value;
+    onlineUsersSelector.appendChild(newOption);
+}
+
+// removes an option from 'onlineUsersSelector'
+function onlineUsersSelectorRemove(value) {
+    let oldOption = document.querySelector("option[value='" + value + "']");
+    if (oldOption !== null) oldOption.remove();
+}
+
+// focus 'chatMessageInput' when user opens the page
+chatMessageInput.focus();
+
+// submit if the user presses the enter key
+chatMessageInput.onkeyup = function(e) {
+    if (e.keyCode === 13) {  // enter key
+        chatMessageSend.click();
+    }
+};
+
+chatMessageSend.onclick = function() {
+    if (chatMessageInput.value.length === 0) return;
+    chatSocket.send(JSON.stringify({
+        "message": chatMessageInput.value,
+    }));
+    chatMessageInput.value = "";
+};
+
+// chat/static/room.js
+
+let chatSocket = null;
+
+function connect() {
+    chatSocket = new WebSocket("ws://" + window.location.host + "/ws/chat/" + roomName + "/");
+
+    chatSocket.onopen = function(e) {
+        console.log("Successfully connected to the WebSocket.");
+    }
+
+    chatSocket.onclose = function(e) {
+        console.log("WebSocket connection closed unexpectedly. Trying to reconnect in 2s...");
+        setTimeout(function() {
+            console.log("Reconnecting...");
+            connect();
+        }, 2000);
+    };
+
+    chatSocket.onmessage = function(e) {
+        const data = JSON.parse(e.data);
+        console.log(data);
+
+        switch (data.type) {
+            case "chat_message":
+                chatLog.value += data.user + ": " + data.message + "\n";  // new
+                break;
+            case "user_list":
+                for (let i = 0; i < data.users.length; i++) {
+                    onlineUsersSelectorAdd(data.users[i]);
+                }
+                break;
+            case "users_messages":
+                for (let i = 0; i < data.messages.length; i++) {
+                    chatLog.value += data.messages[i] + "\n";
+                }
+                break;
+            case "user_join":
+                chatLog.value += data.user + " joined the room.\n";
+                onlineUsersSelectorAdd(data.user);
+                break;
+            case "user_leave":
+                chatLog.value += data.user + " left the room.\n";
+                onlineUsersSelectorRemove(data.user);
+                break;
+            default:
+                console.error("Unknown message type!");
+                break;
+        }
+
+        // scroll 'chatLog' to the bottom
+        chatLog.scrollTop = chatLog.scrollHeight;
+    };
+
+    chatSocket.onerror = function(err) {
+        console.log("WebSocket encountered an error: " + err.message);
+        console.log("Closing the socket.");
+        chatSocket.close();
+    }
+}
+connect();
```

Comment:
The function and event handlers are too long. Breaking these down into smaller functions (e.g., one to handle incoming messages, another to manage the user list, etc.) can improve readability and maintainability.

Reply:
Done, thanks